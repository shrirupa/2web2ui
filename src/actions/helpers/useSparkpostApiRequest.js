import { useState, useEffect } from 'react';
import { sparkpost as sparkpostAxios } from 'src/helpers/axiosInstances';
import store from 'src/store';
import _ from 'lodash';

function useSparkpostApiRequest(options) {
  const [response, setResponse] = useState({});

  // Force effect to rerun by using a toggled boolean as a dependant
  const [retryToggle, setRetryToggle] = useState(false);
  const retry = () => setRetryToggle(!retryToggle);

  const { auth } = store.getState();
  const httpOptions = { ...options };

  if (auth.loggedIn) {
    _.set(httpOptions, 'headers.Authorization', auth.token);
  }

  useEffect(() => {
    setResponse({ pending: true, error: null });

    sparkpostAxios(httpOptions).then((res) => {
      const results = _.get(res, 'data.results', res.data);
      setResponse({ pending: false, results });
    }).catch((error) => {
      setResponse({ pending: false, error: error });
    });
  }, [retryToggle]);

  return { ...response, retry };
}

export default useSparkpostApiRequest;
