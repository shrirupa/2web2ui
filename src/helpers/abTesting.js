import _ from 'lodash';

/**
 * Formats edit mode's redux-form values to be passed on to the API
 */
export const formatFormValues = ({ default_template, variants, dates, ...rest }) => {
  let values = _.clone(rest);

  if (values.test_mode === 'learning') {
    values = _.omit(values, 'confidence_level');
  } else {
    values.confidence_level = parseFloat(values.confidence_level);
  }

  if (values.audience_selection === 'percent') {
    values = _.omit(values, 'total_sample_size');
    delete default_template.sample_size;
    variants = _.map(variants, (variant) => _.omit(variant, 'sample_size'));
  } else {
    values.total_sample_size = calculateTotalSampleSize({ default_template, variants });
    delete default_template.percent;
    variants = _.map(variants, (variant) => _.omit(variant, 'percent'));
  }

  // Cast stringified values to numbers
  default_template[values.audience_selection] = parseFloat(default_template[values.audience_selection]);
  _.forEach(variants, (variant) => variant[values.audience_selection] = parseFloat(variant[values.audience_selection]));
  values.engagement_timeout = parseInt(values.engagement_timeout, 10);

  return {
    ...values,
    start_time: dates.from,
    end_time: dates.to,
    default_template: reduceTemplateObject(default_template),
    variants: variants.map(reduceTemplateObject)
  };
};

/**
 * Formats template object (default template and variants) from redux-form to be passed on to the API
 * The template typeahead returns the full template object, we only need its ID
 */
export const reduceTemplateObject = ({ template_object, ...rest } = {}) => ({
  template_id: _.get(template_object, 'id'),
  ...rest
});

/**
 * Formats template object (default template and variants) from the API to be passed on to redux-form
 * The template typeahead expects full template objects
 */
export const findTemplateObject = (templates, { template_id, ...rest } = {}) => ({
  template_object: _.find(templates, ({ id }) => id === template_id),
  ...rest
});

/**
 * Calculates the total_sample_size of a test based upon the sample_size of default_template & variants
 */
export const calculateTotalSampleSize = ({ default_template, variants }) =>
  _.reduce(
    variants,
    (sum, variant) => sum + parseInt(variant.sample_size, 10),
    parseInt(default_template.sample_size, 10)
  );

export const formatDraftOptionalValues = (values) => {
  _.remove(values.variants, (variant) => 'percent' in variant && !_.isFinite(variant.percent));
  _.remove(values.variants, (variant) => 'sample_size' in variant && !_.isFinite(variant.sample_size));

  if (values.variants.length === 0) {
    delete values.variants;
  }

  if ('percent' in values.default_template && !_.isFinite(values.default_template.percent)) {
    values.default_template.percent = 100;
  }

  if ('sample_size' in values.default_template && !_.isFinite(values.default_template.sample_size)) {
    values.default_template.sample_size = 1;
  }

  if ('total_sample_size' in values && (values.total_sample_size === 0 || isNaN(values.total_sample_size))) {
    values.total_sample_size = 1;
  }

  return values;
};
