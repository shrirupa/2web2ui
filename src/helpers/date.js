import moment from 'moment';
import config from 'src/config';
import { roundBoundaries } from './metrics';
import { FORMATS } from 'src/constants';

export const relativeDateOptions = [
  { value: 'hour', label: 'Last Hour' },
  { value: 'day', label: 'Last 24 Hours' },
  { value: '7days', label: 'Last 7 Days' },
  { value: '10days', label: 'Last 10 Days' },
  { value: '14days', label: 'Last 14 Days' },
  { value: '30days', label: 'Last 30 Days' },
  { value: '90days', label: 'Last 90 Days' },
  { value: 'custom', label: 'Custom' }
];

export const relativeDateOptionsIndexed = relativeDateOptions.reduce((result, { value, label }) => {
  result[value] = label;
  return result;
}, {});

export const getRelativeDateOptions = (ranges) => relativeDateOptions.filter((item) => ranges.includes(item.value));

/**
 * Takes a date string and returns the end of that day (11:59PM)
 *
 * If preventFuture is true and the given day IS the current day,
 * it returns the current time, i.e. the closest to the end
 * of the day without going into the future.
 *
 * @param {String} date - date string to base date on
 * @return {Date}
 */
export function getEndOfDay(date, { preventFuture } = {}) {
  const now = moment();
  const end = new Date(date);

  if (preventFuture && now.isSame(end, 'day')) {
    return now.toDate();
  }

  end.setHours(23, 59, 59, 999);
  return end;
}

export function getStartOfDay(date) {
  const start = new Date(date);
  start.setHours(0, 0, 0, 0);
  return start;
}

export function getNextHour(date) {
  const roundedDate = new Date(date);
  const now = new Date(Date.now());
  roundedDate.setHours(now.getHours() + 1, 0, 0, 0);
  return roundedDate;
}

export function isSameDate(a, b) {
  return (a instanceof Date) && (b instanceof Date) && (a.getTime() === b.getTime());
}

/**
 * Use native JS methods to get user's local
 * timezone, if those methods are present
 *
 * Defaults to UTC otherwise
 */
export function getLocalTimezone() {
  // eslint-disable-next-line new-cap
  if (typeof Intl === 'undefined' || typeof Intl.DateTimeFormat !== 'function') {
    return 'UTC';
  }

  // eslint-disable-next-line new-cap
  const format = Intl.DateTimeFormat();

  if (typeof format.resolvedOptions !== 'function') {
    return 'UTC';
  }

  return format.resolvedOptions().timeZone;
}

export function getRelativeDates(range, { now = moment().toDate(), roundToPrecision = true } = {}) {
  let preciseFrom;

  switch (range) {
    case 'hour':
      preciseFrom = moment(now).subtract(1, 'hour').toDate();
      break;

    case 'day':
      preciseFrom = moment(now).subtract(1, 'day').toDate();
      break;

    case '7days':
      preciseFrom = moment(now).subtract(7, 'day').toDate();
      break;

    case '10days':
      preciseFrom = moment(now).subtract(10, 'day').toDate();
      break;

    case '14days':
      preciseFrom = moment(now).subtract(14, 'day').toDate();
      break;

    case '30days':
      preciseFrom = moment(now).subtract(30, 'day').toDate();
      break;

    case '90days':
      preciseFrom = moment(now).subtract(90, 'day').toDate();
      break;

    case 'custom':
      return { relativeRange: range };

    default: {
      if (range) {
        throw new Error(`Tried to calculate a relative date range with an invalid range value: ${range}`);
      }
      return {};
    }
  }

  if (roundToPrecision) {
    const { to, from } = roundBoundaries(preciseFrom, now);
    return { to: to.toDate(), from: from.toDate(), relativeRange: range };
  }

  return { to: now, from: preciseFrom, relativeRange: range };
}

export function getDuration(dates, unit = 'hours') {
  const { from, to } = dates;
  return moment(to).diff(moment(from), unit);
}

export function formatDate(date, FORMAT = config.dateFormat) {
  return moment(date).format(FORMAT);
}

export function formatTime(time, FORMAT = config.timeFormat) {
  return moment(time).format(FORMAT);
}

export function formatDateTime(datetime) {
  return `${formatDate(datetime)}, ${formatTime(datetime)}`;
}

export const formatInputDate = (date) => moment(date).format(FORMATS.SHORT_DATE);
export const formatInputTime = (time) => moment(time).format(FORMATS.TIME);
export const parseDate = (str) => moment(str, FORMATS.INPUT_DATES, true);
export const parseTime = (str) => moment(str, FORMATS.INPUT_TIMES, true);
export const parseDatetime = (...args) => moment(args.join(' '), FORMATS.INPUT_DATETIMES, true);

export const fillByDate = ({ dataSet, fill = {}, from, now, relativeRange, to } = {}) => {
  let dates = { from, to };

  if (relativeRange && relativeRange !== 'custom') {
    dates = getRelativeDates(relativeRange, { now });
  }

  const orderedData = dataSet.sort((a, b) => new Date(a.date) - new Date(b.date));
  let filledDataSet = [];

  for (let time = moment(dates.from), index = 0; time.isBefore(moment(dates.to)); time.add(1, 'day')) {
    const data = orderedData[index];
    const fillData = { ...fill, date: formatInputDate(time) };

    if (!data || data.date !== fillData.date) {
      filledDataSet = [...filledDataSet, fillData];
    } else {
      filledDataSet = [...filledDataSet, data];
      index++;
    }
  }

  return filledDataSet;
};

/**
 * Generates 3 dates based off relative range
 * Returns first date, middle date, and 1 day before to
 */
export function getDateTicks(relativeRange) {
  const { from, to } = getRelativeDates(relativeRange, { now: moment().subtract(1, 'day') });
  const diff = moment(to).diff(from, 'days');
  const middle = moment(from).add(diff / 2, 'days');

  return [
    formatInputDate(from),
    formatInputDate(middle),
    formatInputDate(to)
  ];
}
