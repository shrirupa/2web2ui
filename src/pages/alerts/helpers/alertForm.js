import {
  FILTERS_FRIENDLY_NAMES,
  METRICS,
  OPERATOR_FRIENDLY_NAMES,
  REALTIME_FILTERS,
  SIGNALS_FILTERS,
  SOURCE_FRIENDLY_NAMES,
  RECOMMENDED_METRIC_VALUE,
} from '../constants/formConstants';

export const getOptionsFromMap = (items, friendlyNameMap) =>
  items.map(item => ({ label: friendlyNameMap[item], value: item }));

const realtimeMetricsSpec = metric => ({
  hasFilters: true,
  filterType: 'multi',
  filterOptions: getOptionsFromMap(REALTIME_FILTERS, FILTERS_FRIENDLY_NAMES),
  sourceOptions: getOptionsFromMap(['raw'], SOURCE_FRIENDLY_NAMES),
  defaultRecommendedValue: RECOMMENDED_METRIC_VALUE[metric].raw.gt,
  defaultFieldValues: [
    { fieldName: 'source', fieldValue: 'raw' },
    { fieldName: 'operator', fieldValue: 'gt' },
  ],
});

const metricToFormSpecMap = {
  monthly_sending_limit: {
    hasFilters: false,
    filterType: null,
    filterOptions: [],
    sourceOptions: getOptionsFromMap(['raw'], SOURCE_FRIENDLY_NAMES),
    defaultRecommendedValue: RECOMMENDED_METRIC_VALUE['monthly_sending_limit'].raw.gt,
    defaultFieldValues: [
      { fieldName: 'subaccounts', fieldValue: [] },
      { fieldName: 'source', fieldValue: 'raw' },
      { fieldName: 'operator', fieldValue: 'gt' },
    ],
  },
  block_bounce_rate: realtimeMetricsSpec('block_bounce_rate'),
  hard_bounce_rate: realtimeMetricsSpec('hard_bounce_rate'),
  soft_bounce_rate: realtimeMetricsSpec('soft_bounce_rate'),
  health_score: {
    hasFilters: true,
    filterType: 'single',
    filterOptions: getOptionsFromMap(SIGNALS_FILTERS, FILTERS_FRIENDLY_NAMES),
    sourceOptions: getOptionsFromMap(
      ['raw', 'week_over_week', 'day_over_day'],
      SOURCE_FRIENDLY_NAMES,
    ),
    defaultRecommendedValue: RECOMMENDED_METRIC_VALUE['health_score'].raw.lt,
    defaultFieldValues: [
      { fieldName: 'source', fieldValue: 'raw' },
      { fieldName: 'operator', fieldValue: 'lt' },
    ],
  },
  injection_count: {
    hasFilters: false,
    filterType: null,
    filterOptions: [],
    sourceOptions: getOptionsFromMap(['raw'], SOURCE_FRIENDLY_NAMES),
    defaultFieldValues: [
      { fieldName: 'subaccounts', fieldValue: [] },
      { fieldName: 'source', fieldValue: 'raw' },
      { fieldName: 'operator', fieldValue: 'lt' },
      { fieldName: 'value', fieldValue: 100000 },
    ],
  },
  blacklist: {
    hasFilters: true,
    hideEvaluator: true,
    hideSubaccountFilter: true,
    filterType: 'multi',
    filterOptions: getOptionsFromMap(
      ['blacklist_provider', 'blacklist_resource'],
      FILTERS_FRIENDLY_NAMES,
    ),
    sourceOptions: [],
    defaultFieldValues: [{ fieldName: 'subaccounts', fieldValue: [] }],
  },
};

export const getFormSpec = metric => {
  if (metric in METRICS) {
    return metricToFormSpecMap[metric];
  }
  return {};
};

export const getEvaluatorOptions = (metric, source) => {
  const operatorArray =
    metric === 'injection_count'
      ? ['lt']
      : metric === 'health_score' && source === 'raw'
      ? ['lt', 'gt']
      : ['gt'];
  const operatorOptions = getOptionsFromMap(operatorArray, OPERATOR_FRIENDLY_NAMES);

  const suffix =
    ['health_score', 'injection_count'].includes(metric) && source === 'raw' ? '' : '%';

  const getSliderProps = () => {
    const sourceLabels = ['week_over_week', 'day_over_day'];
    if (sourceLabels.includes(source)) {
      return { label: 'Percent Change', precision: 0 };
    } else if (source !== 'raw') {
      return { label: 'Percent Change', precision: 2 };
    } else {
      switch (metric) {
        case 'health_score':
          return { label: 'Score', precision: 1 };
        case 'monthly_sending_limit':
          return { label: 'Percent Used', precision: 0 };
        case 'block_bounce_rate':
        case 'hard_bounce_rate':
        case 'soft_bounce_rate':
          return { label: 'Bounce Percentage Above', precision: 0 };
        case 'injection_count':
          return {
            label: 'Falls Below',
            max: 1000000,
            min: 1000,
            precision: 0,
          };
      }
    }

    return {};
  };

  return { suffix, operatorOptions, sliderProps: getSliderProps() };
};
