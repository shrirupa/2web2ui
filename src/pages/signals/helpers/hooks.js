// JSON.stringify() does not guarantee the serialised order of object members.
export const comparableJson = (obj) => JSON.stringify(Object.keys(obj).sort().map((key) => ({ [key]: obj[key] })));
