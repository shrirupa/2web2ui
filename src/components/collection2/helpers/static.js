import { objectSortMatch } from 'src/helpers/sortMatch';
const objectValuesToString = (keys) => (item) => (keys || Object.keys(item)).map((key) => item[key]).join(' ');

/*
 * Filter a fixed rowset using objectSortMatch()
 */
const filter = (rows, pattern = '', { keyMap, itemToStringKeys, matchThreshold } = {}) => {
  if (!pattern) {
    return { rows, rowCount: rows.length };
  }
  const filtered = objectSortMatch({
    items: rows,
    pattern,
    getter: objectValuesToString(itemToStringKeys),
    keyMap,
    matchThreshold
  });
  return {
    rows: filtered,
    rowCount: filtered.length
  };
};

export const filteringModel = (rows, options) => ({ pattern } = {}) => filter(rows, pattern, options);

/*
 * Construct a single "page" of results from a fixed rowset
 */
const paginate = (rows, currentPage = 1, perPage = 25) => {
  const currentIndex = (currentPage - 1) * perPage;
  return {
    rows: rows.slice(currentIndex, currentIndex + perPage),
    rowCount: rows.length
  };
};

export const paginatingModel = (rows) => ({ currentPage, perPage } = {}) => paginate(rows, currentPage, perPage);

/*
 * Create a filter/pagination function for a fixed list of records. This is meant for use
 * as a fetchRows implementation for a Collection.
 */
export const fullModel = (rows, { keyMap = {}, itemToStringKeys, matchThreshold } = {}) =>
  ({ pattern, currentPage, perPage } = {}) => {
    const { rows: filtered, rowCount } = filter(rows, pattern, { keyMap, itemToStringKeys, matchThreshold });
    return {
      rows: paginate(filtered, currentPage, perPage).rows,
      rowCount
    };
  };
