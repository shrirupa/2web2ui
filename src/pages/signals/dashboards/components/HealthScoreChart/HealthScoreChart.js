import React, { useState, Fragment } from 'react';
import { connect } from 'react-redux';
import { getCaretProps, getDoD } from 'src/helpers/signals';
import { selectCurrentHealthScoreDashboard } from 'src/selectors/signals';
import { Panel } from '@sparkpost/matchbox';
import BarChart from '../../../components/charts/barchart/BarChart';
import TooltipMetric from '../../../components/charts/tooltip/TooltipMetric';
import { PanelLoading } from 'src/components';
import Callout from 'src/components/callout';
import MetricDisplay from '../MetricDisplay/MetricDisplay';
import { formatNumber, roundToPlaces } from 'src/helpers/units';
import thresholds from '../../../constants/healthScoreThresholds';
import { formatDate, getDateTicks } from 'src/helpers/date';
import moment from 'moment';
import _ from 'lodash';
import styles from './HealthScoreChart.module.scss';

export function HealthScoreChart(props) {
  const [hoveredDate, setHovered] = useState(props.defaultHovered);

  function handleDateHover(props) {
    setHovered(props.date);
  }

  const { history = [], loading, error, filters } = props;
  const noData = (history) ? !_.some(getHealthScores()) : true;
  const lastItem = _.last(history) || {};
  const selectedDate = _.get(lastItem, 'date', '');

  if (loading) {
    return <PanelLoading minHeight='407px' />;
  }

  if (error) {
    return (
      <Panel sectioned>
        <div className={styles.Content}>
          <div className={styles.ErrorMessage}>
            <Callout title='Unable to Load Data' height='auto' children={error.message} />
          </div>
        </div>
      </Panel>
    );
  }

  function getHealthScores() {
    return _.map(history, (entry) => entry.health_score);
  }

  function getXAxisProps() {
    const xTicks = getDateTicks(filters);
    return {
      ticks: xTicks,
      tickFormatter: (tick) => moment(tick).format('M/D')
    };
  }

  function getTotalInjectionProps() {
    const injectionEntries = _.filter(history, (entry) => entry.injections);
    const injectionTotals = _.map(injectionEntries, (entry) => entry.injections);
    const sumInjectionTotals = _.sum(injectionTotals);
    return { value: _.isNil(sumInjectionTotals) ? 'n/a' : formatNumber(sumInjectionTotals) };
  }

  function getHoverInjectionProps() {
    const currentEntry = _.find(history, ['date', hoveredDate]);
    const value = (currentEntry) ? currentEntry.injections : null;
    return { value: _.isNil(value) ? 'n/a' : formatNumber(value) };
  }

  function getHoverDoDProps() {
    const previousDay = moment(hoveredDate).subtract(1, 'day').format('YYYY-MM-DD');
    const currentEntry = _.find(history, ['date', hoveredDate]);
    const prevEntry = _.find(history, ['date', previousDay]);
    const currentScore = (currentEntry) ? currentEntry.health_score : null;
    const prevScore = (prevEntry) ? prevEntry.health_score : null;
    const value = getDoD(currentScore, prevScore);
    return { value: _.isNil(value) ? 'n/a' : `${value}%`, ...getCaretProps(value) };
  }

  function getGap() {
    return history.length > 15 ? 0.2 : 1;
  }

  function getMin() {
    const min = _.min(getHealthScores());
    return { value: _.isNil(min) ? 'n/a' : min };
  }

  function getMax() {
    const max = _.max(getHealthScores());
    return { value: _.isNil(max) ? 'n/a' : max };
  }

  return (
    <Panel sectioned>
      <div className={styles.Content}>
        <h2 className={styles.Header}>
          {formatDate(filters.from)} â€“ {formatDate(filters.to)}
        </h2>
        {noData && <div><Callout height='auto'>Health Scores Not Available</Callout></div>}
        {!noData && (
          <Fragment>
            <BarChart
              gap={getGap()}
              onMouseOver={handleDateHover}
              onMouseOut={() => setHovered({})}
              selected={selectedDate}
              hovered={hoveredDate}
              timeSeries={history}
              tooltipContent={({ payload = {}}) => (payload.ranking) &&
                (<TooltipMetric
                  label='Health Score'
                  color={thresholds[payload.ranking].color}
                  value={`${roundToPlaces(payload.health_score, 1)}`}
                />)
              }
              yAxisRefLines={[
                { y: 80, stroke: thresholds.good.color, strokeWidth: 1 },
                { y: 55, stroke: thresholds.danger.color, strokeWidth: 1 }
              ]}
              yKey='health_score'
              xAxisProps={getXAxisProps()}
              yAxisProps={{ ticks: [0,55,80,100]}}
            />
          </Fragment>
        )}
        <div className={styles.Metrics}>
          <MetricDisplay label='Total Injections' {...getTotalInjectionProps()} />
          <MetricDisplay label='High' {...getMax()} />
          <MetricDisplay label='Low' {...getMin()} />
          {(typeof hoveredDate === 'string') && (<Fragment>
            <div className={styles.Divider} />
            <div className={styles.Injections}>
              <MetricDisplay label='Injections' {...getHoverInjectionProps()} />
            </div>
            <div className={styles.DoDChange}>
              <MetricDisplay label='DoD Change' {...getHoverDoDProps()} />
            </div>
          </Fragment>
          )}
        </div>
      </div>
    </Panel>
  );
}

const mapStateToProps = (state) => ({
  filters: state.signalOptions,
  ...selectCurrentHealthScoreDashboard(state)
});

export default connect(mapStateToProps, {})(HealthScoreChart);
