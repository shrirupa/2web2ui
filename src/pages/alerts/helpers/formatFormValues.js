import _ from 'lodash';
import { NOTIFICATION_CHANNELS } from '../constants/formConstants';
import { getFormSpec } from './alertForm';
import { multilineStringToArray } from 'src/helpers/string';

export default function formatFormValues(values) {
  const keysToOmit = [
    'value',
    'source',
    'operator',
    'sending_ip',
    'mailbox_provider',
    'blacklist_provider',
    'blacklist_resource',
    'sending_domain',
    'single_filter',
    ...NOTIFICATION_CHANNELS,
  ];
  const { metric, value, source, operator, single_filter, subaccounts } = values;

  const any_subaccount = subaccounts.length === 1 && subaccounts[0] === -2 ? true : undefined;

  //If subaccounts is empty, default to '-1'
  if (subaccounts.length === 0) {
    subaccounts.push(-1);
  }

  /*
  This gets the correct filters for the different metric types.
  First it looks up the form spec for that metric to get the form type.
  For single: returns empty array if no facet is selected, else return only that filter
  For multi: Loops through the realtime filters and adds the filters that have values.
  By default, return an empty array.
   */
  const { filterOptions, filterType, hideEvaluator } = getFormSpec(metric);

  const getFiltersMap = {
    single: () => (single_filter.filter_type === 'none' ? [] : [single_filter]),
    multi: () =>
      filterOptions
        .map(
          ({ value: filter_type }) =>
            values[filter_type].length > 0 && {
              filter_type,
              filter_values: values[filter_type],
            },
        )
        .filter(Boolean),
    default: () => [],
  };
  const filters = (getFiltersMap[filterType] || getFiltersMap.default)();

  const channels = {};
  const emails = (values.emails || '').trim();
  const slack = (values.slack || '').trim();
  const webhook = (values.webhook || '').trim();

  if (emails) {
    channels.emails = multilineStringToArray(emails);
  }

  if (slack) {
    channels.slack = { target: slack };
  }

  if (webhook) {
    channels.webhook = { target: webhook };
  }

  const keysToChange = {
    subaccounts: any_subaccount ? undefined : subaccounts,
    any_subaccount,
    filters,
    threshold_evaluator: hideEvaluator ? {} : { operator, source, value },
    channels,
  };

  return _.omit({ ...values, ...keysToChange }, keysToOmit);
}
