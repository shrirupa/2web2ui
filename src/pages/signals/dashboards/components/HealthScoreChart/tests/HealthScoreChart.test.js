import { shallow } from 'enzyme';
import React from 'react';
import { HealthScoreChart } from '../HealthScoreChart';
import * as dateMock from 'src/helpers/date';

jest.mock('src/helpers/date');

describe('Signals Health Score Chart', () => {
  let props; let subject;

  beforeEach(() => {
    props = {
      loading: false,
      error: null,
      data: [{
        sid: -1,
        current_health_score: 88,
        WoW: -5, current_DoD: 5,
        history: [{
          date: '2019-03-24',
          health_score: 75,
          ranking: 'warning'
        },{
          date: '2019-03-25',
          health_score: 96,
          ranking: 'good'
        },{
          date: '2019-03-26',
          health_score: 23,
          ranking: 'danger'
        },{
          date: '2019-03-25',
          health_score: null,
          ranking: null
        }]
      }],
      filters: {
        relativeRange: '90days'
      }
    };

    dateMock.getDateTicks.mockImplementation(() => [1,2]);
    subject = (options = {}) => shallow(
      <HealthScoreChart {...props} {...options} />
    );
  });

  it('renders happy path correctly', () => {
    expect(subject(props)).toMatchSnapshot();
  });

  it('renders loading correctly', () => {
    const wrapper = subject({ loading: true });
    expect(wrapper.find('PanelLoading')).toExist();
  });

  it('renders error correctly', () => {
    const wrapper = subject({ error: { message: 'mock error' }});
    expect(wrapper.find('Callout')).toMatchSnapshot();
  });

  it('renders no account level data callout correctly', () => {
    props.data = [];
    const wrapper = subject(props);
    expect(wrapper.find('Callout')).toMatchSnapshot();
  });

  it('renders yesterday as selected date if one is missing', () => {
    props.data[0].history = [];
    expect(subject(props)).toMatchSnapshot();
  });
});
