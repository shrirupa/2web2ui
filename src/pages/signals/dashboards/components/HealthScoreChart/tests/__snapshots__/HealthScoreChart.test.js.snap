// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signals Health Score Chart renders error correctly 1`] = `
<Callout
  height="auto"
  title="Unable to Load Data"
>
  mock error
</Callout>
`;

exports[`Signals Health Score Chart renders happy path correctly 1`] = `
<Panel
  sectioned={true}
>
  <div
    className="Content"
  >
    <h2
      className="Header"
    >
      Dec 1 2014
       â€“ 
      Feb 1 2015
    </h2>
    <BarChart
      activeFill="#22838A"
      fill="#B3ECEF"
      gap={1}
      height={250}
      margin={
        Object {
          "bottom": 5,
          "left": 18,
          "right": 0,
          "top": 12,
        }
      }
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      selected="2019-03-27"
      timeSeries={
        Array [
          Object {
            "date": "2019-03-24",
            "health_score": 75,
            "injections": 75000000000,
            "ranking": "warning",
          },
          Object {
            "date": "2019-03-25",
            "health_score": 96,
            "injections": 75000000000,
            "ranking": "good",
          },
          Object {
            "date": "2019-03-26",
            "health_score": 23,
            "injections": 75000000000,
            "ranking": "danger",
          },
          Object {
            "date": "2019-03-27",
            "health_score": null,
            "ranking": null,
          },
        ]
      }
      tooltipContent={[Function]}
      width="99%"
      xAxisProps={
        Object {
          "tickFormatter": [Function],
          "ticks": Array [
            "2018-12-26",
            "2019-02-09",
            "2019-03-26",
          ],
        }
      }
      xAxisRefLines={Array []}
      xKey="date"
      yAxisProps={
        Object {
          "ticks": Array [
            0,
            55,
            80,
            100,
          ],
        }
      }
      yAxisRefLines={
        Array [
          Object {
            "stroke": "#2CC995",
            "strokeWidth": 1,
            "y": 80,
          },
          Object {
            "stroke": "#FF594D",
            "strokeWidth": 1,
            "y": 55,
          },
        ]
      }
      yKey="health_score"
      yRange={
        Array [
          "auto",
          "auto",
        ]
      }
    />
    <div
      className="Metrics"
    >
      <MetricDisplay
        label="Total Injections"
        value="225B"
      />
      <MetricDisplay
        label="High"
        value={96}
      />
      <MetricDisplay
        label="Low"
        value={23}
      />
    </div>
  </div>
</Panel>
`;

exports[`Signals Health Score Chart renders metric displays 1`] = `
<MetricDisplay
  label="Total Injections"
  value="225B"
/>
`;

exports[`Signals Health Score Chart renders metric displays 2`] = `
<MetricDisplay
  label="High"
  value={96}
/>
`;

exports[`Signals Health Score Chart renders no account level data callout correctly 1`] = `
<Callout
  height="auto"
>
  Health Scores Not Available
</Callout>
`;
